pilot/ARCHITECTURE.md

```diff
@@ -0,0 +1,163 @@
+
+As a result, they do not go through `PushContext`, and changes do not trigger a `PushContext` recomputation. Instead, the current state is incrementally computed based on events from `ServiceDiscovery`.
+
+#### Conclusion
+Overall, the high level config ingestion flow:
+```mermaid
+graph TD
+	sd(Service Discovery)
+	cs(ConfigStore)
+	ep(Endpoints)
+	pc(PushContext)
+	sd-->pc
+	cs-->pc
+	sd-->ep
+```
+### Config Translation
+Config Translation turns the above inputs into the actual types consumed by the connected XDS clients (typically Envoy). This is done by `Generators`, which register a function to build a given type. For example, there is a `RouteGenerator` responsible for building `Routes`. Along with the core Envoy XDS types, there are a few custom Istio types, such as our `NameTable` type used for DNS, as well as debug interfaces.
+
+`Generators` get as input the `Proxy` (a representation of the current client), the current `PushContext` snapshot, and a list of config updates that caused the change.
+
+The `Proxy` as an input parameter is important, and a major distinction from some other XDS implementations. We are not able to statically translate inputs to XDS without per-client information. For example, we rely on the client's labels to determine the set of policies applied. While this is necessary to implement Istio's APIs, it does limit performance substantially.
+#### Caching
+Config translation typically takes the overwhelming majority of Istiod's resource usage. In particular, protobuf encoding. As a result, caching has been introduced, storing the already encoded `protobuf.Any` for a given resource.
+
+This caching depends on declaring all inputs to the given generator as part of the cache key. This is extremely error-prone, as there is nothing preventing generators from consuming inputs that are *not* part of the key. When this happens, different clients will non-deterministically get incorrect configuration. This type of bug has historically resulted in CVEs.
+
+There are a few ways to prevent these:
+* Only pass in to the generation logic the cache key itself, so no other unaccounted inputs can be used. Unfortunately, this has not been done for any generators today.
+* Be very, very careful.
+* The cache has a builtin test, enabled with `UNSAFE_PILOT_ENABLE_RUNTIME_ASSERTIONS=true`, that runs in CI. This will panic if any key is written to with a different value.
+#### Partial Computations
+Along with caching, partial computations are a critical performance optimization to ensure that we do not need to build (or send) every resource to every proxy on every change. This is discussed more in the Config Serving section.
+
+### Config Serving
+Config serving is the layer that actually accepts proxy clients, connected over bidirectional gRPC streams, and serve them the required configuration.
+
+We will have two triggers for sending config - requests and pushes.
```

Comment:
Is this model designed to prefer/biased to favor one or the other?

i.e. does istiod try to proactively push everything, and client requests are viewed as sort of a "cache miss" fallback?

Reply:
We favor requests over pushes actually, since those are blocking envoy. 

We cannot proactively push new things, as envoy will drop them on the floor if its not subscribed. So you could maybe coincidentally push a resource it wanted but generally its not reliable

Reply:
Also 'on connect' is critical for startup latency - while pushes for existing clients can be safely delayed ( for config). 
The goal was to intentionally stage/delay to post-connect push.

Worth mentioning that Envoy is no longer the XDS client - but Istio-agent, where we can push some things proactively. Or ztunnel, proxyless.